src/ast/pass_manager.h

```diff
@@ -1,64 +1,371 @@
+    // simply wrap this in something that returns OK on the user's behalf.
+    static_assert((std::is_reference_v<Inputs> && ...),
+                  "all arguments must be references");
+    if constexpr (std::is_same_v<Return, void>) {
+      std::function<Result<OK>(Inputs...)> wrap(
+          [fn](Inputs... inputs) -> Result<OK> {
+            fn(inputs...);
+            return OK();
+          });
+      return createImpl(name, wrap);
+    } else {
+      std::function<Result<Return>(Inputs...)> wrap(
+          [fn](Inputs... inputs) -> Result<Return> { return fn(inputs...); });
+      return createImpl(name, wrap);
+    }
+  }
+  template <typename Return, typename... Inputs>
+  static Pass createImpl(const std::string &name,
+                         std::function<Result<Return>(Inputs &...)> fn)
+  {
+    std::vector<int> inputs = { PassContext::TypeId<Inputs>::type_id()... };
+    std::vector<int> outputs;
+    if constexpr (!std::is_same_v<OK, Return>) {
+      // Note that the Return value may be a reference, so we resolve the
+      // `type_id` of the underlying type. The Input types are not references,
+      // because of the way the template is constructed.
+      outputs = { PassContext::TypeId<std::decay_t<Return>>::type_id() };
+    }
 
-public:
-  std::string name;
+    return Pass(name,
+                std::move(inputs),
+                std::move(outputs),
+                [fn](PassContext &ctx) -> Result<> {
+                  // Extract a tuple of all input parameters, apply the function
+                  // and set the result.
+                  auto args = std::make_tuple(std::ref(ctx.get<Inputs>())...);
+                  auto result = std::apply(fn, args);
+                  if (!result) {
+                    return result.takeError();
+                  }
+                  if constexpr (std::is_same_v<Return, OK>) {
+                    return OK();
+                  } else {
+                    if constexpr (std::is_reference_v<Return>) {
+                      // Add the external reference.
+                      ctx.put(*result);
+                    } else {
+                      // Just move the result.
+                      ctx.put(std::move(*result));
+                    }
+                    return OK();
+                  }
+                });
+  }
+
+  // N.B. external construction is done via `create`.
+  using Fn = std::function<Result<>(PassContext &ctx)>;
+  Pass(std::string name,
+       std::vector<int> &&inputs,
+       std::vector<int> &&outputs,
+       Fn fn)
+      : name_(std::move(name)),
+        inputs_(std::move(inputs)),
+        outputs_(std::move(outputs)),
+        fn_(std::move(fn))
+  {
+  }
+
+  std::string name_;
+  std::vector<int> inputs_;
+  std::vector<int> outputs_;
+  Fn fn_;
 };
 
+// PassManager hold passes in sequence.
+//
+// The PassManager is responsible for verifying the inputs and outputs of the
```

Comment:
This is now said a third time - maybe just a summary at the top of PassManager would be better.

Reply:
Dropped two.